Update Discord user status for multiplayer/playlists

LeNitrous opened the pull request:
Some multiplayer screens aren't using the correct activities. This affects `DiscordRichPresence`.
`MultiplayerMatchSongSelect` originally did not set `OsuScreen.InitialActivity`. I've used `UserActivity.InLobby` since it is still under a multiplayer lobby.
![image](https://user-images.githubusercontent.com/20495991/129435084-4c3b4085-798e-4d7f-8897-237b3866a34d.png)

`MultiplayerPlayer` inherited `Player` thus uses `UserActivity.SoloGame` making `UserActivity.MultiplayerGame` unused. I've also expanded it to "`<ruleset playing verb>` with others"
![image](https://user-images.githubusercontent.com/20495991/129435150-3bcbb709-c9f4-49f7-88b3-0052e6e8ebe3.png)

LeNitrous committed:
update `InitialActivity` on multiplayer `Player` and `SongSelect`

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerMatchSongSelect.cs:
```diff
@@ -19,8 +20,13 @@ public class MultiplayerMatchSongSelect : OnlinePlaySongSelect
+
         private LoadingLayer loadingLayer;
 
+        protected override UserActivity InitialActivity => new UserActivity.InLobby(room);
```
This is not done for `PlaylistsSongSelect` which seems inconsistent.

In fact this should *probably* be moved to `OnlinePlaySongSelect`.

LeNitrous replied:
Would be good if we changed `UserActivity.InLobby.Status` to "In a lobby" instead of "In a multiplayer lobby" perhaps?

bdach replied:
I suppose so.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:
```diff
@@ -25,6 +26,8 @@ public class MultiplayerPlayer : RoomSubmittingPlayer
     {
         protected override bool PauseOnFocusLost => false;
 
+        protected override UserActivity InitialActivity => new UserActivity.MultiplayerGame(Beatmap.Value.BeatmapInfo, Ruleset.Value);
```
Similarly, this should probably live in `RoomSubmittingPlayer`.

LeNitrous replied:
I think this is fine as-is since you aren't technically playing with others live compared to multiplayer.

bdach replied:
On the other side, it's not really "solo" either, is it? That's how it is right now.

bdach commented on osu.Game/Users/UserActivity.cs:
```diff
@@ -25,9 +25,14 @@ public class ChoosingBeatmap : UserActivity
+            {
+            }
+
+            public override string Status => $@"{base.Status} with others";
```
This won't work for rulesets that don't override `PlayingVerb`.

https://github.com/ppy/osu/blob/78ac092147d994aa8a832584a66ad4df42c45c1b/osu.Game/Rulesets/Ruleset.cs#L226-L229

It's not a huge deal for now as only the built-in ones are permitted and those all override it, but the potential of having this return `Playing solo with others` indicates that this may not be the optimal way to do this.

(Also, as an aside, this is a bit of an interesting one wrt localisation - should the status returned to discord be localised according to the user's selected language, or not?)

LeNitrous replied:
Should `Ruleset.PlayingVerb` be changed to "Playing" then? It is expected that many custom rulesets may want to override `PlayingVerb`.

Also, I don't think localization is necessary as many other games that use Rich Presence (and Spotify too if that counts) use English.

bdach replied:
I guess that could work? No strong opinion from me.

frenzibyte replied:
Might as well make it `abstract` and avoid arbitrary strings?

bdach replied:
Breaking ruleset API change, so I'm not sure I agree.

LeNitrous committed:
apply suggestions

LeNitrous committed:
trim whitespace

peppy commented on osu.Game/Users/UserActivity.cs:
```diff
@@ -25,9 +25,14 @@ public class ChoosingBeatmap : UserActivity
         }
 
-        public class MultiplayerGame : UserActivity
+        public class MultiplayerGame : SoloGame
```
not 100% about this inheritance... feels a bit weird.

bdach replied:
glossed over this but fair point. should probably be `InGame : UserActivity`, `InSoloGame : InGame`, `InMultiplayerGame : InGame`

LeNitrous commented:
I've also noticed the lobby name doesn't update when it does. I suppose its still within the scope of this PR right?

LeNitrous commented:
I've looked into attempting updating the lobby name and while it seems to be in-scope of this PR, it may add complexity overall. Maybe as a separate one instead perchance.

LeNitrous committed:
apply suggestions

- make `UserActivity.InGame` and derive that to `InSoloGame` and `InMultiplayerGame`
- rename `SoloGame` to `InSoloGame`
- rename `MultiplayerGame` to `InMultiplayerGame`

LeNitrous committed:
apply code inspection fixes

frenzibyte commented on osu.Game/Rulesets/Ruleset.cs:
```diff
@@ -224,9 +224,9 @@ public PerformanceCalculator CreatePerformanceCalculator(WorkingBeatmap beatmap,
 
         /// <summary>
-        /// The playing verb to be shown in the <see cref="UserActivity.SoloGame.Status"/>.
+        /// The playing verb to be shown in the <see cref="UserActivity.InGame.Status"/>.
```
```suggestion
        /// The playing verb to be shown in the <see cref="UserActivity.InGame"/> activities.
```

frenzibyte commented on osu.Game/Users/UserActivity.cs:
```diff
@@ -25,36 +25,49 @@ public class ChoosingBeatmap : UserActivity
-        }
-
-        public class Editing : UserActivity
+        public class InGame : UserActivity
```
Isn't this suppsoed to be `abstract`?

LeNitrous replied:
Code Style rules required me to make it `protected`. Unless that is ignorable.

frenzibyte replied:
It's accessed outside, no? it can't be `protected`

LeNitrous committed:
apply suggestions

peppy commented on osu.Game/Screens/Play/RoomSubmittingPlayer.cs:
```diff
@@ -19,6 +20,8 @@ public abstract class RoomSubmittingPlayer : SubmittingPlayer
 
         protected readonly PlaylistItem PlaylistItem;
 
+        protected override UserActivity InitialActivity => new UserActivity.InMultiplayerGame(Beatmap.Value.BeatmapInfo, Ruleset.Value);
```
Note that this is covering both playlists and multiplayer currently. I'm not sure if you want to expand on this to split the two apart, but it may be best, since playlists is not really "playing with others" or a `MultiplayerGame`.

LeNitrous replied:
We had this discussion with @bdach already: https://github.com/ppy/osu/pull/14284#discussion_r688935948
Should I take up on your suggestion then?

bdach replied:
I don't really have any objections to splitting. As long as both are covered somehow.

LeNitrous replied:
I'm unsure of what status text to use in playlists if `InPlaylistGame` is going to be introduced. Currently this is what I have in mind instead:

Solo — `InSoloGame`
Multi — `InMultiplayerGame`
Playlists - `InSoloGame` (because this is technically a solo play)

bdach replied:
If we're splitting three-ways then might as well have playlists separate and not inherit solo game, as there may be more info you might want to show (like the playlist name or something).

LeNitrous committed:
split multiplayer and playlist activity

smoogipoo committed:
Merge branch 'master' into activity-on-multiplayer-screens

smoogipoo committed:
Refactor to remove resolved dependency

smoogipoo committed:
Fix test compile errors

bdach committed:
Merge branch 'master' into activity-on-multiplayer-screens

bdach commented:
I've resolved a (rather trivial) merge conflict here.
